src/main/java/io/supertokens/bulkimport/BulkImportUserUtils.java

```diff
@@ -0,0 +1,551 @@
+                    true, JsonArray.class, errors, " for a user role.");
+
+            role = validateAndNormaliseUserRole(role, allUserRoles, errors);
+            List<String> normalisedTenantIds = validateAndNormaliseTenantIds(main, appIdentifier, jsonTenantIds, errors,
+                    " for a user role.");
+
+            if (role != null && normalisedTenantIds != null) {
+                userRoles.add(new UserRole(role, normalisedTenantIds));
+            }
+        }
+        return userRoles;
+    }
+
+    private static List<TotpDevice> getParsedTotpDevices(JsonObject userData, List<String> errors) {
+        JsonArray jsonTotpDevices = parseAndValidateFieldType(userData, "totpDevices", ValueType.ARRAY_OF_OBJECT, false,
+                JsonArray.class, errors, ".");
+
+        if (jsonTotpDevices == null) {
+            return null;
+        }
+
+        List<TotpDevice> totpDevices = new ArrayList<>();
+        for (JsonElement jsonTotpDeviceEl : jsonTotpDevices) {
+            JsonObject jsonTotpDevice = jsonTotpDeviceEl.getAsJsonObject();
+
+            String secretKey = parseAndValidateFieldType(jsonTotpDevice, "secretKey", ValueType.STRING, true,
+                    String.class, errors, " for a totp device.");
+            Integer period = parseAndValidateFieldType(jsonTotpDevice, "period", ValueType.INTEGER, false,
+                    Integer.class, errors, " for a totp device.");
+            Integer skew = parseAndValidateFieldType(jsonTotpDevice, "skew", ValueType.INTEGER, false, Integer.class,
+                    errors, " for a totp device.");
+            String deviceName = parseAndValidateFieldType(jsonTotpDevice, "deviceName", ValueType.STRING, false,
+                    String.class, errors, " for a totp device.");
+
+            secretKey = validateAndNormaliseTotpSecretKey(secretKey, errors);
+            period = validateAndNormaliseTotpPeriod(period, errors);
+            skew = validateAndNormaliseTotpSkew(skew, errors);
+            deviceName = validateAndNormaliseTotpDeviceName(deviceName, errors);
+
+            if (secretKey != null && period != null && skew != null) {
+                totpDevices.add(new TotpDevice(secretKey, period, skew, deviceName));
+            }
+        }
+        return totpDevices;
+    }
+
+    private static List<LoginMethod> getParsedLoginMethods(Main main, AppIdentifier appIdentifier, JsonObject userData,
+            List<String> errors)
+            throws StorageQueryException, TenantOrAppNotFoundException {
+        JsonArray jsonLoginMethods = parseAndValidateFieldType(userData, "loginMethods", ValueType.ARRAY_OF_OBJECT,
+                true, JsonArray.class, errors, ".");
+
+        if (jsonLoginMethods == null) {
+            return new ArrayList<>();
+        }
+
+        if (jsonLoginMethods.size() == 0) {
+            errors.add("At least one loginMethod is required.");
+            return new ArrayList<>();
```

Comment:
if the user has added totp device info, there should be MFA enabled

Reply:
done